<a href="https://colab.research.google.com/github/froschi95/HMLE2021/blob/main/ML_Code_Club_wk1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# DATA WRANGLING OF CELLPHONE DATASET 

## The dataset was obtained from [Gapminder](https://www.gapminder.org/)


In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

In [2]:
cell_phones = pd.read_csv('https://raw.githubusercontent.com/zalihat/cell_phone_dataset/master/cell_phones.csv', error_bad_lines=False)
population = pd.read_csv('https://raw.githubusercontent.com/zalihat/cell_phone_dataset/master/population1.csv', error_bad_lines=False)
geo_codes = pd.read_csv('https://raw.githubusercontent.com/open-numbers/ddf--gapminder--systema_globalis/master/ddf--entities--geo--country.csv', error_bad_lines=False)

In [3]:
cell_phones.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11640 entries, 0 to 11639
Data columns (total 3 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Country            11640 non-null  object 
 1   year               11640 non-null  int64  
 2   cell_phones_total  11640 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 272.9+ KB


In [4]:
population.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 58695 entries, 0 to 58694
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Country     58695 non-null  object
 1   year        58695 non-null  int64 
 2   population  58695 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 1.3+ MB


In [5]:
geo_codes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 273 entries, 0 to 272
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   country                 273 non-null    object 
 1   g77_and_oecd_countries  259 non-null    object 
 2   income_3groups          218 non-null    object 
 3   income_groups           218 non-null    object 
 4   is--country             273 non-null    bool   
 5   iso3166_1_alpha2        246 non-null    object 
 6   iso3166_1_alpha3        247 non-null    object 
 7   iso3166_1_numeric       247 non-null    float64
 8   iso3166_2               1 non-null      object 
 9   landlocked              255 non-null    object 
 10  latitude                241 non-null    float64
 11  longitude               241 non-null    float64
 12  main_religion_2008      216 non-null    object 
 13  name                    273 non-null    object 
 14  un_sdg_ldc              249 non-null    ob

In [6]:
geo_codes.isnull().sum()

country                     0
g77_and_oecd_countries     14
income_3groups             55
income_groups              55
is--country                 0
iso3166_1_alpha2           27
iso3166_1_alpha3           26
iso3166_1_numeric          26
iso3166_2                 272
landlocked                 18
latitude                   32
longitude                  32
main_religion_2008         57
name                        0
un_sdg_ldc                 24
un_sdg_region              25
un_state                    0
unicef_region              78
unicode_region_subtag      25
world_4region               2
world_6region              13
dtype: int64

In [7]:
geo_codes.head()

country g77_and_oecd_countries  ... world_4region        world_6region
0    abkh                 others  ...        europe  europe_central_asia
1     abw                 others  ...      americas              america
2     afg                    g77  ...          asia           south_asia
3     ago                    g77  ...        africa   sub_saharan_africa
4     aia                 others  ...      americas              america

[5 rows x 21 columns]

In [8]:
population.head()

Country  year  population
0  Afghanistan  1800     3280000
1  Afghanistan  1801     3280000
2  Afghanistan  1802     3280000
3  Afghanistan  1803     3280000
4  Afghanistan  1804     3280000

## JOIN DATA

In [9]:
data = cell_phones.merge(population, how='inner', on=['Country', 'year'])
data.head()

Country  year  cell_phones_total  population
0  Afghanistan  1960                0.0     9000000
1  Afghanistan  1961                0.0     9170000
2  Afghanistan  1962                0.0     9350000
3  Afghanistan  1963                0.0     9540000
4  Afghanistan  1964                0.0     9740000

In [10]:
data.shape

(11640, 4)

In [11]:
geo_codes.rename(columns={'country': 'geo', 'name':'Country'}, inplace=True)

In [12]:
geo_codes.head()

geo g77_and_oecd_countries  ... world_4region        world_6region
0  abkh                 others  ...        europe  europe_central_asia
1   abw                 others  ...      americas              america
2   afg                    g77  ...          asia           south_asia
3   ago                    g77  ...        africa   sub_saharan_africa
4   aia                 others  ...      americas              america

[5 rows x 21 columns]

In [13]:
merged_data = data.merge(geo_codes[['geo', 'Country']], how='inner', on='Country')
merged_data.head()
# print(merged_data.shape)

Country  year  cell_phones_total  population  geo
0  Afghanistan  1960                0.0     9000000  afg
1  Afghanistan  1961                0.0     9170000  afg
2  Afghanistan  1962                0.0     9350000  afg
3  Afghanistan  1963                0.0     9540000  afg
4  Afghanistan  1964                0.0     9740000  afg

## MAKE FEATURES

In [14]:
merged_data['cellphones_per_person'] = merged_data['cell_phones_total'] / merged_data['population']
merged_data.head()

Country  year  cell_phones_total  population  geo  cellphones_per_person
0  Afghanistan  1960                0.0     9000000  afg                    0.0
1  Afghanistan  1961                0.0     9170000  afg                    0.0
2  Afghanistan  1962                0.0     9350000  afg                    0.0
3  Afghanistan  1963                0.0     9540000  afg                    0.0
4  Afghanistan  1964                0.0     9740000  afg                    0.0

In [15]:
# Assert

merged_data[(merged_data['year'] == 2017) & (merged_data['Country'] == 'United States')]


Country  year  ...  geo  cellphones_per_person
11157  United States  2017  ...  usa               1.230769

[1 rows x 6 columns]

In [16]:
merged_data['geo'] = merged_data['geo'].str.upper()
merged_data.head()

Country  year  cell_phones_total  population  geo  cellphones_per_person
0  Afghanistan  1960                0.0     9000000  AFG                    0.0
1  Afghanistan  1961                0.0     9170000  AFG                    0.0
2  Afghanistan  1962                0.0     9350000  AFG                    0.0
3  Afghanistan  1963                0.0     9540000  AFG                    0.0
4  Afghanistan  1964                0.0     9740000  AFG                    0.0

## PROCESS DATA

In [17]:
merged_data.describe(include='all')

Country          year  ...    geo  cellphones_per_person
count       11640  11640.000000  ...  11640           11640.000000
unique        194           NaN  ...    194                    NaN
top     Argentina           NaN  ...    IND                    NaN
freq           60           NaN  ...     60                    NaN
mean          NaN   1989.500000  ...    NaN               0.235536
std           NaN     17.318846  ...    NaN               0.433303
min           NaN   1960.000000  ...    NaN               0.000000
25%           NaN   1974.750000  ...    NaN               0.000000
50%           NaN   1989.500000  ...    NaN               0.000000
75%           NaN   2004.250000  ...    NaN               0.233667
max           NaN   2019.000000  ...    NaN               2.126068

[11 rows x 6 columns]

The result above shows that there are 194 unique values for `'Country'` and `'geo'`.
The most common country in the dataset is 'Lesotho', occuring 60 times. Also, the most common geocode in the data is 'CHE' occuring 60 times.

For the Numeric Columms:The records were obtained from 1960 to 2019

Population ranges from 27 milliion to 1.4 billion.

Total cellphones ranges from 7875722 to 1.73 billion

Cellphones per individual ranges from 0 to 2


In [18]:
pd.set_option('display.float_format', str)

sorted_countries_by_cellphone_total = merged_data[merged_data['year'] == 2017][['Country', 'year', 'cell_phones_total']].sort_values(by='cell_phones_total', ascending=False)
sorted_countries_by_cellphone_total.head()

Country  year  cell_phones_total
2157           China  2017       1470000000.0
4617           India  2017       1170000000.0
4677       Indonesia  2017        435000000.0
11157  United States  2017        400000000.0
8577          Russia  2017        227000000.0

In [19]:
sorted_countries_by_cellphone_total.dtypes

Country               object
year                   int64
cell_phones_total    float64
dtype: object

In [20]:
# merged_data[(merged_data['year'] == 2017) & (merged_data['Country'] == 'China')]